### ISSUE TITLE
Disable build cache in execution time

### ISSUE THREAD

**Issue by mockitoguy** [gradle/gradle#7769]
### Problem

Today we can turn off Build Cache for a task using "output.cacheIf()" or "output.doNotCacheIf()". However, we can only do it before the task started execution. We would like to disable build cache for given task _after_ it has executed (for example with "doLast" or TaskExecutionListener).

### Use case

Some of our verification tasks find issues that require build failure. However, we don't want to fail individual verification task. Rather, we want to fail later, after all specific verification tasks have executed, when we can show a custom report showing all failures, across many verification tasks, across many submodules. Example:

```
* What went wrong:
Execution failed for task ':reportAggregatedFailures'.
> Tasks reported failures (http://go/code-quality):
    :demo:test - 1 failure - file:///.../build/demo/reports/tests/test/index.html
    :demo:forkedCheckstyleMainProcessCheckstyle - 1 issue - file:///.../build/demo/reports/checkstyle/forked-main.html
```

How is this use case relevant to build cache? If the task's result is SUCCESS but it actually reported failures that require the build failure, the build cache stores the result of this task. If the task is flaky then we are hosed - we cannot easily re-run the build to get successful build. This is painful to the teams who keep trying to re-run the build and get the same failure. Of course we keep working on the flaky tests problem but we also need to get the team unblocked.

We talked about this use case with @hansd and @mark-vieira last week when they visited LinkedIn.

### Continue on failure?

The "deferred failure report" use case is similar to "continue on failure" Gradle feature, usually used via command line flag. However COF does not have a couple of capabilities we need:
 - we want custom, super clean report of the failures, where we can sort, deduplicate and sanitize
 - COF triggers too many task executions (slower builds, unrelated failures reported). There are legit reasons where plugins declare "mustRunAfter" instead of "dependsOn" to offer more flexibility to the user. That way of hooking up the tasks makes COF less useful.

### Summary

Gradle team, what do you think about making build cache control more flexible, and allowing to disable "build cache store" for given task, after it has executed?

**Comment by mockitoguy** [comment]
Hey Gradle team, any thoughts about this idea?

**Comment by mockitoguy** [comment]
Today we had to turn off our custom failure aggregation for all builds, all teams. I'd love to provide great failure reports and the build cache to all our teams.

**Comment by DPUkyle** [comment]
+1 It would be nice to have an API to decide during/after a task executes to decide if we want to "keep" the results or dispose of them.

The `checkstyle` example above is a good one. A failure in the `checkstyle` task should _eventually_ cause a build failure, but it's not so severe as to justify terminating build execution at the conclusion of that task.

**Comment by hansd** [comment]
See also https://github.com/gradle/build-cache/issues/255

**Comment by pioterj** [comment]
@mockitoguy Sorry for not following up in a timely fashion. Just to confirm - is this still an issue for you or have you found some workaround in the meantime?

**Comment by IronLung7** [comment]
Any update for this issue?

**Comment by lptr** [comment]
Let's reopen this if the need arises.

**Comment by staktrace** [comment]
@lptr Can you please reopen this issue? It's not clear why you closed this, as it appears to still be an issue. There doesn't appear to be a way to prevent a task from getting cached once execution has started and if the task produces a success result.

**Comment by staktrace** [comment]
Ok, thanks. TLDR: I no longer need gradle to support this API (although it would have simplified my code) so I'm not going to open a new issue.

For context the use case I was having trouble with was the following:
- During the task execution I do a network call to fetch some data, and then use that data in some computation
- If the network call succeeds, I want the task output to be cached
- If the network call fails, I want the task to still succeed (graceful handling of network failure), but want the output to NOT be cached, so that subsequent executions of the task try again

After thinking about this more I ended up splitting this into two tasks. One task just does the network call and saves the result to a task output. If the network call fails, it saves an empty task output. The second task does the computation, using the first task's output as an input. If that first task's output is empty (indicating the network call failed), then I mark the second task as not `upToDate` so that the second task is not cached. This works because the `upToDate` condition for the second task is evaluated after execution of the first task, so it can effectively "know" if the network call passed or failed. The second task additionally does its own network fetch if the first task failed, and the first task gets cached for 24 hours. This provided a close-enough outcome for what I needed, and if others run into a similar problem I hope this description helps them.
